File: sdk/python/feast/feature_table.py
# Copyright 2020 The Feast Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, List, MutableMapping, Optional, Union

import yaml
from google.protobuf import json_format
from google.protobuf.duration_pb2 import Duration
from google.protobuf.json_format import MessageToDict, MessageToJson
from google.protobuf.timestamp_pb2 import Timestamp

from feast.core.FeatureTable_pb2 import FeatureTable as FeatureTableProto
from feast.core.FeatureTable_pb2 import FeatureTableMeta as FeatureTableMetaProto
from feast.core.FeatureTable_pb2 import FeatureTableSpec as FeatureTableSpecProto
from feast.data_source import (
    BigQuerySource,
    DataSource,
    FileSource,
    KafkaSource,
    KinesisSource,
)
from feast.feature import Feature
from feast.loaders import yaml as feast_yaml
from feast.value_type import ValueType


class FeatureTable:
    """
    Represents a collection of features and associated metadata.
    """

    def __init__(
        self,
        name: str,
        entities: List[str],
        features: List[Feature],
        batch_source: Union[BigQuerySource, FileSource] = None,
        stream_source: Optional[Union[KafkaSource, KinesisSource]] = None,
        max_age: Optional[Duration] = None,
        labels: Optional[MutableMapping[str, str]] = None,
    ):
        self._name = name
        self._entities = entities
        self._features = features
        self._batch_source = batch_source
        self._stream_source = stream_source
        if labels is None:
            self._labels = dict()  # type: MutableMapping[str, str]
        else:
            self._labels = labels

        self._max_age = max_age
        self._created_timestamp: Optional[Timestamp] = None
        self._last_updated_timestamp: Optional[Timestamp] = None

    def __str__(self):
        return str(MessageToJson(self.to_proto()))

    def __eq__(self, other):
        if not isinstance(other, FeatureTable):
            raise TypeError(
                "Comparisons should only involve FeatureTable class objects."
            )

        if (
            self.labels != other.labels
            or self.name != other.name
            or self.max_age != other.max_age
        ):
            return False

        if sorted(self.entities) != sorted(other.entities):
            return False
        if sorted(self.features) != sorted(other.features):
            return False
        if self.batch_source != other.batch_source:
            return False
        if self.stream_source != other.stream_source:
            return False

        return True

    @property
    def name(self):
        """
        Returns the name of this feature table
        """
        return self._name

    @name.setter
    def name(self, name: str):
        """
        Sets the name of this feature table
        """
        self._name = name

    @property
    def entities(self):
        """
        Returns the entities of this feature table
        """
        return self._entities

    @entities.setter
    def entities(self, entities: List[str]):
        """
        Sets the entities of this feature table
        """
        self._entities = entities

    @property
    def features(self):
        """
        Returns the features of this feature table
        """
        return self._features

    @features.setter
    def features(self, features: List[Feature]):
        """
        Sets the features of this feature table
        """
        self._features = features

    @property
    def batch_source(self):
        """
        Returns the batch source of this feature table
        """
        return self._batch_source

    @batch_source.setter
    def batch_source(self, batch_source: Union[BigQuerySource, FileSource]):
        """
        Sets the batch source of this feature table
        """
        self._batch_source = batch_source

    @property
    def stream_source(self):
        """
        Returns the stream source of this feature table
        """
        return self._stream_source

    @stream_source.setter
    def stream_source(self, stream_source: Union[KafkaSource, KinesisSource]):
        """
        Sets the stream source of this feature table
        """
        self._stream_source = stream_source

    @property
    def max_age(self):
        """
        Returns the maximum age of this feature table. This is the total maximum
        amount of staleness that will be allowed during feature retrieval for
        each specific feature that is looked up.
        """
        return self._max_age

    @max_age.setter
    def max_age(self, max_age: Duration):
        """
        Set the maximum age for this feature table
        """
        self._max_age = max_age

    @property
    def labels(self):
        """
        Returns the labels of this feature table. This is the user defined metadata
        defined as a dictionary.
        """
        return self._labels

    @labels.setter
    def labels(self, labels: MutableMapping[str, str]):
        """
        Set the labels for this feature table
        """
        self._labels = labels

    @property
    def created_timestamp(self):
        """
        Returns the created_timestamp of this feature table
        """
        return self._created_timestamp

    @property
    def last_updated_timestamp(self):
        """
        Returns the last_updated_timestamp of this feature table
        """
        return self._last_updated_timestamp

    def add_feature(self, feature: Feature):
        """
        Adds a new feature to the feature table.
        """
        self.features.append(feature)

    def is_valid(self):
        """
        Validates the state of a feature table locally. Raises an exception
        if feature table is invalid.
        """

        if not self.name:
            raise ValueError("No name found in feature table.")

        if not self.entities:
            raise ValueError("No entities found in feature table {self.name}.")

    @classmethod
    def from_yaml(cls, yml: str):
        """
        Creates a feature table from a YAML string body or a file path

        Args:
            yml: Either a file path containing a yaml file or a YAML string

        Returns:
            Returns a FeatureTable object based on the YAML file
        """

        return cls.from_dict(feast_yaml.yaml_loader(yml, load_single=True))

    @classmethod
    def from_dict(cls, ft_dict):
        """
        Creates a feature table from a dict

        Args:
            ft_dict: A dict representation of a feature table

        Returns:
            Returns a FeatureTable object based on the feature table dict
        """

        feature_table_proto = json_format.ParseDict(
            ft_dict, FeatureTableProto(), ignore_unknown_fields=True
        )

        return cls.from_proto(feature_table_proto)

    @classmethod
    def from_proto(cls, feature_table_proto: FeatureTableProto):
        """
        Creates a feature table from a protobuf representation of a feature table

        Args:
            feature_table_proto: A protobuf representation of a feature table

        Returns:
            Returns a FeatureTableProto object based on the feature table protobuf
        """

        feature_table = cls(
            name=feature_table_proto.spec.name,
            entities=[entity for entity in feature_table_proto.spec.entities],
            features=[
                Feature(
                    name=feature.name,
                    dtype=ValueType(feature.value_type),
                    labels=feature.labels,
                )
                for feature in feature_table_proto.spec.features
            ],
            labels=feature_table_proto.spec.labels,
            max_age=(
                None
                if feature_table_proto.spec.max_age.seconds == 0
                and feature_table_proto.spec.max_age.nanos == 0
                else feature_table_proto.spec.max_age
            ),
            batch_source=DataSource.from_proto(feature_table_proto.spec.batch_source),
            stream_source=(
                None
                if not feature_table_proto.spec.stream_source.ByteSize()
                else DataSource.from_proto(feature_table_proto.spec.stream_source)
            ),
        )

        feature_table._created_timestamp = feature_table_proto.meta.created_timestamp

        return feature_table

    def to_proto(self) -> FeatureTableProto:
        """
        Converts an feature table object to its protobuf representation

        Returns:
            FeatureTableProto protobuf
        """

        meta = FeatureTableMetaProto(
            created_timestamp=self.created_timestamp,
            last_updated_timestamp=self.last_updated_timestamp,
        )

        spec = FeatureTableSpecProto(
            name=self.name,
            entities=self.entities,
            features=[
                feature.to_proto() if type(feature) == Feature else feature
                for feature in self.features
            ],
            labels=self.labels,
            max_age=self.max_age,
            batch_source=(
                self.batch_source.to_proto()
                if issubclass(type(self.batch_source), DataSource)
                else self.batch_source
            ),
            stream_source=(
                self.stream_source.to_proto()
                if issubclass(type(self.stream_source), DataSource)
                else self.stream_source
            ),
        )

        return FeatureTableProto(spec=spec, meta=meta)

    def to_spec_proto(self) -> FeatureTableSpecProto:
        """
        Converts an FeatureTableProto object to its protobuf representation.
        Used when passing FeatureTableSpecProto object to Feast request.

        Returns:
            FeatureTableSpecProto protobuf
        """

        spec = FeatureTableSpecProto(
            name=self.name,
            entities=self.entities,
            features=[
                feature.to_proto() if type(feature) == Feature else feature
                for feature in self.features
            ],
            labels=self.labels,
            max_age=self.max_age,
            batch_source=(
                self.batch_source.to_proto()
                if issubclass(type(self.batch_source), DataSource)
                else self.batch_source
            ),
            stream_source=(
                self.stream_source.to_proto()
                if issubclass(type(self.stream_source), DataSource)
                else self.stream_source
            ),
        )

        return spec

    def to_dict(self) -> Dict:
        """
        Converts feature table to dict

        :return: Dictionary object representation of feature table
        """
        feature_table_dict = MessageToDict(self.to_proto())

        # Remove meta when empty for more readable exports
        if feature_table_dict["meta"] == {}:
            del feature_table_dict["meta"]

        return feature_table_dict

    def to_yaml(self):
        """
        Converts a feature table to a YAML string.

        :return: Feature table string returned in YAML format
        """
        feature_table_dict = self.to_dict()
        return yaml.dump(feature_table_dict, allow_unicode=True, sort_keys=False)

    def _update_from_feature_table(self, feature_table):
        """
        Deep replaces one feature table with another

        Args:
            feature_table: Feature table to use as a source of configuration
        """

        self.name = feature_table.name
        self.entities = feature_table.entities
        self.features = feature_table.features
        self.labels = feature_table.labels
        self.max_age = feature_table.max_age
        self.batch_source = feature_table.batch_source
        self.stream_source = feature_table.stream_source
        self._created_timestamp = feature_table.created_timestamp
        self._last_updated_timestamp = feature_table.last_updated_timestamp

    def __repr__(self):
        return f"FeatureTable <{self.name}>"

File: java/core/src/main/java/feast/core/model/FeatureTable.java
/*
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2018-2020 The Feast Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package feast.core.model;

import static feast.common.models.FeatureV2.getFeatureStringRef;

import com.google.common.hash.Hashing;
import com.google.protobuf.Duration;
import com.google.protobuf.Timestamp;
import feast.core.dao.EntityRepository;
import feast.core.util.TypeConversion;
import feast.proto.core.DataSourceProto;
import feast.proto.core.FeatureProto.FeatureSpecV2;
import feast.proto.core.FeatureTableProto;
import feast.proto.core.FeatureTableProto.FeatureTableSpec;
import feast.proto.serving.ServingAPIProto;
import java.util.*;
import java.util.stream.Collectors;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;

@Getter
@Entity
@Setter(AccessLevel.PRIVATE)
@Table(
    name = "feature_tables",
    uniqueConstraints = @UniqueConstraint(columnNames = {"name", "project_name"}))
public class FeatureTable extends AbstractTimestampEntity {

  @Id @GeneratedValue private long id;

  // Name of Feature Table
  @Column(name = "name", nullable = false)
  private String name;

  // Name of the Project that this FeatureTable belongs to
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "project_name")
  private Project project;

  // Features defined in this Feature Table
  @OneToMany(
      mappedBy = "featureTable",
      cascade = CascadeType.ALL,
      fetch = FetchType.EAGER,
      orphanRemoval = true)
  private Set<FeatureV2> features;

  // Entites to associate the features defined in this FeatureTable with
  @ManyToMany(fetch = FetchType.EAGER)
  @JoinTable(
      name = "feature_tables_entities_v2",
      joinColumns = @JoinColumn(name = "feature_table_id"),
      inverseJoinColumns = @JoinColumn(name = "entity_v2_id"))
  private Set<EntityV2> entities;

  // User defined metadata labels serialized as JSON string.
  @Column(name = "labels", columnDefinition = "text")
  private String labelsJSON;

  // Max Age of the Features defined in this Feature Table in seconds
  @Column(name = "max_age", nullable = false)
  private long maxAgeSecs;

  // Streaming DataSource used to obtain data for features from a stream
  @OneToOne(cascade = CascadeType.ALL)
  @JoinColumn(name = "stream_source_id", nullable = true)
  private DataSource streamSource;

  // Batched DataSource used to obtain data for features from a batch of data
  @OneToOne(cascade = CascadeType.ALL)
  @JoinColumn(name = "batch_source_id", nullable = false)
  private DataSource batchSource;

  // Autoincrementing version no. of this FeatureTable.
  // Autoincrements every update made to the FeatureTable.
  @Column(name = "revision", nullable = false)
  private int revision;

  @Column(name = "is_deleted", nullable = false)
  private boolean isDeleted;

  public FeatureTable() {};

  /**
   * Construct FeatureTable from Protobuf spec representation in the given project with entities
   * registered in entity repository.
   *
   * @param projectName the name of the project that the constructed FeatureTable belongs.
   * @param spec the Protobuf spec to construct the Feature from.
   * @param entityRepo {@link EntityRepository} used to resolve entity names.
   * @throws IllegalArgumentException if the Protobuf spec provided is invalid.
   * @return constructed FeatureTable from the given Protobuf spec.
   */
  public static FeatureTable fromProto(
      String projectName, FeatureTableSpec spec, EntityRepository entityRepo) {
    FeatureTable table = new FeatureTable();
    table.setName(spec.getName());
    table.setProject(new Project(projectName));

    Set<FeatureV2> features =
        spec.getFeaturesList().stream()
            .map(featureSpec -> FeatureV2.fromProto(table, featureSpec))
            .collect(Collectors.toSet());
    table.setFeatures(features);

    Set<EntityV2> entities =
        FeatureTable.resolveEntities(
            projectName, spec.getName(), entityRepo, spec.getEntitiesList());
    table.setEntities(entities);

    String labelsJSON = TypeConversion.convertMapToJsonString(spec.getLabelsMap());
    table.setLabelsJSON(labelsJSON);

    table.setMaxAgeSecs(spec.getMaxAge().getSeconds());
    table.setBatchSource(DataSource.fromProto(spec.getBatchSource()));

    // Configure stream source only if set
    if (!spec.getStreamSource().equals(DataSourceProto.DataSource.getDefaultInstance())) {
      table.setStreamSource(DataSource.fromProto(spec.getStreamSource()));
    }

    return table;
  }

  /**
   * Update the FeatureTable from the given Protobuf representation.
   *
   * @param projectName project name
   * @param spec the Protobuf spec to update the FeatureTable from.
   * @param entityRepo repository
   * @throws IllegalArgumentException if the update will make prohibited changes.
   */
  public void updateFromProto(
      String projectName, FeatureTableSpec spec, EntityRepository entityRepo) {
    // Check for prohibited changes made in spec:
    // - Name cannot be changed
    if (!getName().equals(spec.getName())) {
      throw new IllegalArgumentException(
          String.format(
              "Updating the name of a registered FeatureTable is not allowed: %s to %s",
              getName(), spec.getName()));
    }
    // Update Entities if changed
    Set<EntityV2> entities =
        FeatureTable.resolveEntities(
            projectName, spec.getName(), entityRepo, spec.getEntitiesList());
    this.setEntities(entities);

    // Update FeatureTable based on spec
    // Update existing features, create new feature, drop missing features
    Map<String, FeatureV2> existingFeatures =
        getFeatures().stream().collect(Collectors.toMap(FeatureV2::getName, feature -> feature));
    this.features.clear();
    this.features.addAll(
        spec.getFeaturesList().stream()
            .map(
                featureSpec -> {
                  if (!existingFeatures.containsKey(featureSpec.getName())) {
                    // Create new Feature based on spec
                    return FeatureV2.fromProto(this, featureSpec);
                  }
                  // Update existing feature based on spec
                  FeatureV2 feature = existingFeatures.get(featureSpec.getName());
                  feature.updateFromProto(featureSpec);
                  return feature;
                })
            .collect(Collectors.toSet()));

    this.maxAgeSecs = spec.getMaxAge().getSeconds();
    this.labelsJSON = TypeConversion.convertMapToJsonString(spec.getLabelsMap());

    this.batchSource = DataSource.fromProto(spec.getBatchSource());
    if (!spec.getStreamSource().equals(DataSourceProto.DataSource.getDefaultInstance())) {
      this.streamSource = DataSource.fromProto(spec.getStreamSource());
    } else {
      this.streamSource = null;
    }

    // Set isDeleted to false
    this.setDeleted(false);

    // Bump revision no.
    this.revision++;
  }

  /**
   * Convert this Feature Table to its Protobuf representation
   *
   * @return protobuf representation
   */
  public FeatureTableProto.FeatureTable toProto() {
    // Convert field types to Protobuf compatible types
    Timestamp creationTime = TypeConversion.convertTimestamp(getCreated());
    Timestamp updatedTime = TypeConversion.convertTimestamp(getLastUpdated());
    String metadataHashBytes = this.protoHash();

    List<FeatureSpecV2> featureSpecs =
        getFeatures().stream().map(FeatureV2::toProto).collect(Collectors.toList());
    List<String> entityNames =
        getEntities().stream().map(EntityV2::getName).collect(Collectors.toList());
    Map<String, String> labels = TypeConversion.convertJsonStringToMap(getLabelsJSON());

    FeatureTableSpec.Builder spec =
        FeatureTableSpec.newBuilder()
            .setName(getName())
            .setMaxAge(Duration.newBuilder().setSeconds(getMaxAgeSecs()).build())
            .setBatchSource(getBatchSource().toProto())
            .addAllEntities(entityNames)
            .addAllFeatures(featureSpecs)
            .putAllLabels(labels);
    if (getStreamSource() != null) {
      spec.setStreamSource(getStreamSource().toProto());
    }

    return FeatureTableProto.FeatureTable.newBuilder()
        .setMeta(
            FeatureTableProto.FeatureTableMeta.newBuilder()
                .setRevision(getRevision())
                .setCreatedTimestamp(creationTime)
                .setLastUpdatedTimestamp(updatedTime)
                .setHash(metadataHashBytes)
                .build())
        .setSpec(spec.build())
        .build();
  }

  /** Use given entity repository to resolve entity names to entity native objects */
  private static Set<EntityV2> resolveEntities(
      String projectName, String tableName, EntityRepository repo, Collection<String> names) {
    return names.stream()
        .map(
            entityName -> {
              EntityV2 entity = repo.findEntityByNameAndProject_Name(entityName, projectName);
              if (entity == null) {
                throw new IllegalArgumentException(
                    String.format(
                        "Feature Table refers to no existent Entity: (table: %s, entity: %s, project: %s)",
                        tableName, entityName, projectName));
              }
              return entity;
            })
        .collect(Collectors.toSet());
  }

  /**
   * Return a boolean to indicate if FeatureTable contains all specified entities.
   *
   * @param entitiesFilter contain entities that should be attached to the FeatureTable
   * @return boolean True if FeatureTable contains all entities in the entitiesFilter
   */
  public boolean hasAllEntities(List<String> entitiesFilter) {
    Set<String> allEntitiesName =
        this.getEntities().stream().map(entity -> entity.getName()).collect(Collectors.toSet());
    return allEntitiesName.equals(new HashSet<>(entitiesFilter));
  }

  /**
   * Returns a map of Feature references and Features if FeatureTable's Feature contains all labels
   * in the labelsFilter
   *
   * @param labelsFilter contain labels that should be attached to FeatureTable's features
   * @return Map of Feature references and Features
   */
  public Map<String, FeatureV2> getFeaturesByLabels(Map<String, String> labelsFilter) {
    Map<String, FeatureV2> validFeaturesMap;
    List<FeatureV2> validFeatures;
    if (labelsFilter.size() > 0) {
      validFeatures = filterFeaturesByAllLabels(this.getFeatures(), labelsFilter);
      validFeaturesMap = getFeaturesRefToFeaturesMap(validFeatures);
      return validFeaturesMap;
    }
    validFeaturesMap = getFeaturesRefToFeaturesMap(List.copyOf(this.getFeatures()));
    return validFeaturesMap;
  }

  /**
   * Returns map for accessing features using their respective feature reference.
   *
   * @param features List of features to insert to map.
   * @return Map of featureRef:feature.
   */
  private Map<String, FeatureV2> getFeaturesRefToFeaturesMap(List<FeatureV2> features) {
    Map<String, FeatureV2> validFeaturesMap = new HashMap<>();
    for (FeatureV2 feature : features) {
      ServingAPIProto.FeatureReferenceV2 featureRef =
          ServingAPIProto.FeatureReferenceV2.newBuilder()
              .setFeatureTable(this.getName())
              .setName(feature.getName())
              .build();
      validFeaturesMap.put(getFeatureStringRef(featureRef), feature);
    }
    return validFeaturesMap;
  }

  /**
   * Returns a list of Features if FeatureTable's Feature contains all labels in labelsFilter
   *
   * @param features features
   * @param labelsFilter contain labels that should be attached to FeatureTable's features
   * @return List of Features
   */
  public static List<FeatureV2> filterFeaturesByAllLabels(
      Set<FeatureV2> features, Map<String, String> labelsFilter) {
    List<FeatureV2> validFeatures =
        features.stream()
            .filter(feature -> feature.hasAllLabels(labelsFilter))
            .collect(Collectors.toList());

    return validFeatures;
  }

  /**
   * Determine whether a FeatureTable has all the specified labels.
   *
   * @param labelsFilter labels contain key-value mapping for labels attached to the FeatureTable
   * @return boolean True if Entity contains all labels in the labelsFilter
   */
  public boolean hasAllLabels(Map<String, String> labelsFilter) {
    Map<String, String> LabelsMap = this.getLabelsMap();
    for (String key : labelsFilter.keySet()) {
      if (!LabelsMap.containsKey(key) || !LabelsMap.get(key).equals(labelsFilter.get(key))) {
        return false;
      }
    }
    return true;
  }

  public Map<String, String> getLabelsMap() {
    return TypeConversion.convertJsonStringToMap(getLabelsJSON());
  }

  public void delete() {
    this.setDeleted(true);
    this.setRevision(0);
  }

  public String protoHash() {
    List<String> sortedEntities =
        this.getEntities().stream().map(EntityV2::getName).sorted().collect(Collectors.toList());

    List<FeatureSpecV2> sortedFeatureSpecs =
        this.getFeatures().stream()
            .sorted(Comparator.comparing(FeatureV2::getName))
            .map(FeatureV2::toProto)
            .collect(Collectors.toList());

    DataSourceProto.DataSource streamSource = DataSourceProto.DataSource.getDefaultInstance();
    if (getStreamSource() != null) {
      streamSource = getStreamSource().toProto();
    }

    FeatureTableSpec featureTableSpec =
        FeatureTableSpec.newBuilder()
            .addAllEntities(sortedEntities)
            .addAllFeatures(sortedFeatureSpecs)
            .setBatchSource(getBatchSource().toProto())
            .setStreamSource(streamSource)
            .setMaxAge(Duration.newBuilder().setSeconds(getMaxAgeSecs()).build())
            .build();
    return Hashing.murmur3_32().hashBytes(featureTableSpec.toByteArray()).toString();
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        getName(),
        getProject(),
        getFeatures(),
        getEntities(),
        getMaxAgeSecs(),
        getBatchSource(),
        getStreamSource());
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof FeatureTable)) {
      return false;
    }

    FeatureTable other = (FeatureTable) o;

    return getName().equals(other.getName())
        && getProject().equals(other.getProject())
        && getLabelsJSON().equals(other.getLabelsJSON())
        && getFeatures().equals(other.getFeatures())
        && getEntities().equals(other.getEntities())
        && getMaxAgeSecs() == other.getMaxAgeSecs()
        && Optional.ofNullable(getBatchSource()).equals(Optional.ofNullable(other.getBatchSource()))
        && Optional.ofNullable(getStreamSource())
            .equals(Optional.ofNullable(other.getStreamSource()));
  }
}
